### PR TITLE
[FEATURE] Enable configurable custom and default text roles

### PR DESCRIPTION
**Pull request by linawolf**
See https://docutils.sourceforge.io/docs/ref/rst/directives.html#role

### CODE REVIEW

File: packages/guides-restructured-text/src/RestructuredText/Directives/DefaultRoleDirective.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        private readonly TextRoleFactory $textRoleFactory,
+    ) {
+    }
+
+    public function getName(): string
+    {
+        return 'default-role';
+    }
+
+    public function process(
```
Comment (jaapio) [inline packages/guides-restructured-text/src/RestructuredText/Directives/DefaultRoleDirective.php:38]: `processAction` was intended for this type of directives. However I noticed that the base directive never calls this method, and it seems never used. Maybe we should consider to enable this usage, by introducting a new base class, that will provide an method like `processAction`

```
abstract class ActionDirective 
{
   process(): {
      $this->processAction();
      return null;
   }
   
    public function processAction(
        DocumentParserContext $documentParserContext,
        string $variable,
        string $data,
        array $options,
    ): void {
    }
}

File: packages/guides-restructured-text/src/RestructuredText/TextRoles/DefaultTextRoleFactory.php [hunk 2]
```diff
@@ -17,19 +17,37 @@ class DefaultTextRoleFactory implements TextRoleFactory
...
         private array $domains = [],
     ) {
         $this->textRoles = [...$textRoles];
     }
 
-    public function registerTextRole(TextRole $textRoles): void
+    public function registerTextRole(TextRole $textRole): void
     {
-        $this->textRoles[] = $textRoles;
+        $this->textRoles[] = $textRole;
```
Comment (jaapio) [inline packages/guides-restructured-text/src/RestructuredText/TextRoles/DefaultTextRoleFactory.php:29]: Should we not create an name based map to store the text roles? As this is there unique identifier?

**Review by jaapio — CHANGES_REQUESTED** [review]
I could not find this in your changes to that's why I ask it, default text roles set by a document are document bases, aren't they? Otherwise the order of parsing matters and I don't thing we have that order currently. It would also disallow caching of the full documents between runs as the parser would be in a different state. Therefor I think it would make sense to add these textroles somehow in the DocumentParserContext? 

Same applies for document defined or replaced roles, those should only apply to the current document from the point where the role was defined.

File: packages/guides-restructured-text/src/RestructuredText/TextRoles/DefaultTextRoleFactory.php (hunk 2 above)
Comment (linawolf) [reply packages/guides-restructured-text/src/RestructuredText/TextRoles/DefaultTextRoleFactory.php:29]: One testrole can have multiple aliases

File: packages/guides-restructured-text/src/RestructuredText/Directives/DefaultRoleDirective.php (hunk 1 above)
Comment (linawolf) [reply packages/guides-restructured-text/src/RestructuredText/Directives/DefaultRoleDirective.php:38]: Good idea

**Comment by linawolf** [comment]
We usually define textroles in the includes.rst.txt which is included at the beginning of each document. So I supose it is probably defined on document level. will have a look at it

File: packages/guides-restructured-text/src/RestructuredText/Directives/DefaultRoleDirective.php (hunk 1 above)
Comment (linawolf) [reply packages/guides-restructured-text/src/RestructuredText/Directives/DefaultRoleDirective.php]: I introduced the abstract class and extended it

**Comment by linawolf** [comment]
@jaapio Each DocumentParserContext now manages a clone of the TextRoleFactory such that changes made to it by one directive do not apply to another document. Also added a test for this.

**Review by jaapio — APPROVED** [review]
Thanks this looks good!

**Pull request merged**